python-package/xgboost/spark/data.py

```diff
@@ -139,6 +140,76 @@ def append_m(part: pd.DataFrame, name: str, is_valid: bool) -> None:
             else:
                 train_data[name].append(array)
 
+    def append_m_sparse(part: pd.DataFrame, name: str, is_valid: bool) -> None:
+        from scipy.sparse import csr_matrix
+        nonlocal n_features
+
+        if name == alias.data or name in part.columns:
+            if name == alias.data:
+                # variables for constructing csr_matrix
+                csr_indices_list, csr_indptr_list, csr_values_list = [], [0], []
+
+                for vec_type, vec_size_, vec_indices, vec_values in zip(
+                    part.featureVectorType,
+                    part.featureVectorSize,
+                    part.featureVectorIndices,
+                    part.featureVectorValues
+                ):
+                    if vec_type == 0:
+                        # sparse vector
+                        vec_size = int(vec_size_)
+                        csr_indices = vec_indices
+                        csr_values = vec_values
+                    else:
+                        # dense vector
+                        # Note: According to spark ML VectorUDT format,
+                        # when type field is 1, the size field is also empty.
+                        # we need to check the values field to get vector length.
+                        vec_size = len(vec_values)
+                        csr_indices = np.arange(vec_size, dtype=np.int32)
+                        csr_values = vec_values
+
+                    if n_features == 0:
+                        n_features = vec_size
+                    assert n_features == vec_size
+
+                    # remove zero elements from csr_indices / csr_values
```

Comment:
Is this necessary? When `missing` is set to 0, XGBoost can remove those values.

Reply:
I remember the DMatrix if data using csc/csr matrix, then it ignore the missing argument but regard all inactive element in the sparse matrix as missing values.

If so, then keep zero elements and removing them represents 2 different semantic:
Keep these zero means it will be regarded as "zero" value feature,
Remove these zero elements means it will be regarded as missing value.

Is my understanding correct ?

Reply:
The ctor for CSR matrix should be able to handle missing values (but not for the CSC, which would raise a warning). https://github.com/dmlc/xgboost/blob/20d1bba1bb8b3aa0ba3762e99e21117b416a8567/python-package/xgboost/data.py#L35

Reply:
which is a good reminder that I should clear the difference.

Reply:
@trivialfis 
OK, so for missing handling, CSR is the same with dense input (respect "missing" param), but CSC is different (ignore "missing" param and regard inactive elements as missing), right ?
We should document this in DMatrix doc.

Reply:
I will update the CSC implementation instead.

Reply:
@trivialfis what does "updating CSC implementation" mean? so for current implementation, it indeed remove the whole instance if one element in the instance is missing value?